File: 5_semester/Тестирование и верификация программного обеспечения/FormValidator/src/FormValidator.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.List;

public class FormValidator {

    public void validateName(String name) throws FormValidationException {
        if (name == null || name.length() < 2 || name.length() > 20) {
            throw new FormValidationException("Имя должно содержать от 2 до 20 символов");
        }

        if (!Character.isUpperCase(name.charAt(0))) {
            throw new FormValidationException("Первая буква имени должна быть заглавной");
        }

        for (char c : name.toCharArray()) {
            if (!Character.isLetter(c) && c != ' ' && c != '-') {
                throw new FormValidationException("Имя может содержать только буквы, пробелы и дефисы");
            }
        }
    }

    public void validateBirthDate(String birthDate) throws FormValidationException {
        try {
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
            LocalDate date = LocalDate.parse(birthDate, formatter);
            LocalDate minDate = LocalDate.of(1900, 1, 1);
            LocalDate currentDate = LocalDate.now();

            if (date.isBefore(minDate)) {
                throw new FormValidationException("Дата рождения не может быть раньше 01.01.1900");
            }

            if (date.isAfter(currentDate)) {
                throw new FormValidationException("Дата рождения не может быть в будущем");
            }
        } catch (DateTimeParseException e) {
            throw new FormValidationException("Неверный формат даты. Используйте формат dd.MM.yyyy");
        }
    }

    public void validateGender(String gender) throws FormValidationException {
        try {
            Gender.valueOf(gender.toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new FormValidationException("Пол должен быть 'MALE' или 'FEMALE'");
        }
    }

    public void validateWeight(String weight) throws FormValidationException {
        try {
            double weightValue = Double.parseDouble(weight);
            if (weightValue <= 0) {
                throw new FormValidationException("Вес должен быть положительным числом");
            }
        } catch (NumberFormatException e) {
            throw new FormValidationException("Вес должен быть числом");
        }
    }

    public boolean validateForm(FormData formData) {
        List<String> errors = new ArrayList<>();

        try {
            validateName(formData.getName());
        } catch (FormValidationException e) {
            errors.add("Ошибка в имени: " + e.getMessage());
        }

        try {
            validateBirthDate(formData.getBirthDate());
        } catch (FormValidationException e) {
            errors.add("Ошибка в дате рождения: " + e.getMessage());
        }

        try {
            validateGender(formData.getGender());
        } catch (FormValidationException e) {
            errors.add("Ошибка в поле: " + e.getMessage());
        }

        try {
            validateWeight(formData.getWeight());
        } catch (FormValidationException e) {
            errors.add("Ошибка в весе: " + e.getMessage());
        }

        if (!errors.isEmpty()) {
            System.out.println("Обнаружены ошибки валидации:");
            for (String error : errors) {
                System.out.println("- " + error);
            }
            return false;
        }

        System.out.println("Анкета прошла валидацию успешно!");
        return true;
    }
}

File: 5_semester/Разработка приложений на языке Котлин/pract/src/pract_10/task_1/FormValidator.kt
package pract_10.task_1

import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException


class FormValidator {
    fun validateName(name: String) {
        if (name.length < 2 || name.length > 20) {
            throw Exception("Имя должно содержать от 2 до 20 символов")
        }

        if (!name[0].isUpperCase()) {
            throw Exception("Первая буква имени должна быть заглавной")
        }
    }

    fun validateBirthDate(birthDate: String) {
        try {
            val formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy")
            val date = LocalDate.parse(birthDate, formatter)
            val minDate = LocalDate.of(1900, 1, 1)
            val currentDate = LocalDate.now()

            if (date.isBefore(minDate)) {
                throw Exception("Дата рождения не может быть раньше 01.01.1900")
            }

            if (date.isAfter(currentDate)) {
                throw Exception("Дата рождения не может быть в будущем")
            }
        } catch (e: DateTimeParseException) {
            throw Exception("Неверный формат даты. Используйте формат dd.MM.yyyy")
        }
    }

    fun validateGender(gender: String) {
        try {
            Gender.valueOf(gender.uppercase())
        } catch (e: IllegalArgumentException) {
            throw Exception("Пол должен быть 'MALE' или 'FEMALE'")
        }
    }

    fun validateWeight(weight: String) {
        try {
            val weightValue = weight.toDouble()
            if (weightValue <= 0) {
                throw Exception("Вес должен быть положительным числом")
            }
        } catch (e: NumberFormatException) {
            throw Exception("Вес должен быть числом")
        }
    }

    fun validateForm(formData: FormData): Boolean {
        val errors = mutableListOf<String>()

        try {
            validateName(formData.name)
        } catch (e: Exception) {
            errors.add("Ошибка в имени: ${e.message}")
        }

        try {
            validateBirthDate(formData.birthDate)
        } catch (e: Exception) {
            errors.add("Ошибка в дате рождения: ${e.message}")
        }

        try {
            validateGender(formData.gender)
        } catch (e: Exception) {
            errors.add("Ошибка в поле: ${e.message}")
        }

        try {
            validateWeight(formData.weight)
        } catch (e: Exception) {
            errors.add("Ошибка в весе: ${e.message}")
        }

        if (errors.isNotEmpty()) {
            println("Обнаружены ошибки валидации:")
            errors.forEach { println("- $it") }
            return false
        }

        println("Анкета прошла валидацию успешно!")
        return true
    }
}